working/1426-extension-types/feature-specification-implements-rep.md

```diff
@@ -0,0 +1,227 @@
+the `implements` clause of an inline class declaration is used to establish
+a subtype relationship to other inline classes. This allows the given
+inline class to "inherit" member implementations from those other inline
+classes, and it establishes a subtype relationship. For example:
+
+```dart
+inline class A {
+  final int i;
+  A(this.i);
+  int get next => i + 1;
+}
+
+inline class B implements A {
+  final int i;
+  B(this.i);
+}
+
+var b = B(1).next; // OK.
+A a = b; // OK.
+int i = b; // Error: `B` is not assignable to `int`.
+```
+
+Another subtype relationship which can safely be adopted is that the inline
+class is a subtype of its representation type, or any supertype of the
+representation type. This proposal broadens the applicability of the
+`implements` clause to establish that kind of subtype relationship as well:
+
+```dart
+inline class A implements int { // `A` is a subtype of `int`.
+  final int i;
+  A(this.i);
+}
+
+var a = A(2);
+int i = a; // OK.
+List<int> list = <A>[a]; // OK.
+```
+
+This subtype relationship is sound because the run-time value of an
+expression of type `A` is an instance of `int`. 
+
+This subtype relationship may be desirable in the case where there is no
+need to protect an object accessed as an `A` against being accessed as an
+`int`, and it is even considered to be useful that it will readily "become
+an `int` whenever needed".
+
+We need to introduce the _ultimate representation type_ of an inline type
+`V<T1 .. Tk>`: This is the instantiated representation type `R` of `V` if
+that is a type that is not and does not contain any inline
+types. Otherwise, assume that `R` is `V1<U1 .. Us>`, then the ultimate
+representation type of `V` is the ultimate representation type of
+`V1<W1 .. Ws>` where `Wj` is the ultimate representation type of `Uj`, for
+`j` in `1 .. s`. Similarly for function types and other composite types.
+
+*In this document it is assumed that the ultimate representation type
+exists. This is ensured by means of a static check on each inline class
+declaration, as specified in the inline class feature specification.*
+
+The corresponding change to the feature specification is be that the
+following sentence in the section
+[Composing inline classes](https://github.com/dart-lang/language/blob/main/accepted/future-releases/inline-classes/feature-specification.md#composing-inline-classes)
+is adjusted:
+
+> A compile-time error occurs if _V1_ is a type name or a parameterized type which occurs as a superinterface in an inline class declaration _DV_, but _V1_ does not denote an inline type.
+
+It is adjusted to end as follows:
+
+> ... declaration _DV_, unless _V1_ denotes an inline type, or _V1_ is a supertype of the ultimate representation type of _DV_.
+
+
+Moreover, a compile-time error occurs if the implements clause of _DV_
+contains two or more types that are non-inline types. A compile-time error
```

Comment:
No need for the "two or more" rule.
If you have:
```dart
class _A implements I1, I2 { ... }
inline class A implements I1, I2 {
   final _A _this;
}
```
there is no problem.

Reply:
I think that's true. However, I was aiming for a minimal model where we could add features like this later (if the complexity vs. benefit trade-off is OK).

Note also that we could certainly do this:

```dart
class _A implements I1, I2 { ... }
inline class A implements _A {
  final _A _this;
  A(this._this);
}
```

This would contribute the same members to `A` as `I1` and `I2` do, with signature resolution as done in `_A`, if needed, plus possibly some members of `_A`. If we just want to use `implements I1, I2` where `I1` and `I2` are non-inline types that are supertypes of the representation type then we can do it like this:

```dart
class _I1AndI2 implements I1, I2 {...} // Only abstract members, resolving conflicts, if any.
inline class A implements _I1AndI2 {...}
```

This means that `_I1AndI2` provides _exactly_ the members from `I1` and from `I2`, with conflict resolution as needed. 

The remaining issue is that `inline class A implements I1, I2` looks nicer in DartDoc.

Still looks like a future extension that we can return to if needed.

Reply:
Did it anyway: There is now no restriction on the number of non-inline superinterfaces. They need to have a combined member signature in cases where more than one of them specify a member signature with the same name.